File: src/commands/batchCoverageCommand.ml
(*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

open CommandUtils
open Utils_js

let spec =
  {
    CommandSpec.name = "batch-coverage";
    doc = "Shows aggregate coverage information for a group of files or directories ";
    usage =
      Printf.sprintf
        "Usage: %s batch-coverage [OPTION]... [FILE...] \n\ne.g. %s batch-coverage foo.js bar.js baz.js dirname1 dirname2 --show-all \nor   %s batch-coverage --input-file filenames.txt\n"
        CommandUtils.exe_name
        CommandUtils.exe_name
        CommandUtils.exe_name;
    args =
      CommandSpec.ArgSpec.(
        empty
        |> base_flags
        |> connect_flags_no_lazy
        |> json_flags
        |> root_flag
        |> strip_root_flag
        |> wait_for_recheck_flag
        |> flag
             "--input-file"
             string
             ~doc:
               "File containing list of files or directories to compute coverage for, one per line. If -, the list is read from standard input."
        |> flag
             "--show-all"
             no_arg
             ~doc:
               "Whether to output the coverage for all files. If not specified, this command will only print coverage for 50 files. "
        |> flag
             "--show-trust"
             no_arg
             ~doc:"EXPERIMENTAL: Whether to include trust information in output"
        |> anon "FILE..." (list_of string));
  }

let output_results ~root ~strip_root ~json ~pretty ~show_all ~trust stats =
  let strip_root =
    if strip_root then
      Some root
    else
      None
  in
  let percent top bottom =
    if bottom = 0 then
      0.
    else
      float_of_int top /. float_of_int bottom *. 100.
  in
  (* Compute aggregate stats *)
  let (trusted, untrusted, any, empty) =
    Base.List.fold_left
      ~f:
        (fun (acc_trust, acc_untrust, acc_any, acc_empty)
             (_, { Coverage_response.untainted; tainted; empty; uncovered }) ->
        (acc_trust + untainted, acc_untrust + tainted, acc_any + uncovered, acc_empty + empty))
      stats
      ~init:(0, 0, 0, 0)
  in
  let num_files_in_dir = Base.List.length stats in
  let covered = trusted + untrusted in
  let total = covered + any + empty in
  let trusted_percentage = percent trusted total in
  let percentage = percent covered total in
  let file_stats (file_key, { Coverage_response.untainted; tainted; empty; uncovered }) =
    let covered = untainted + tainted in
    let total = covered + uncovered + empty in
    let percentage = percent covered total in
    let trusted_percentage =
      if trust then
        Some (percent untainted total)
      else
        None
    in
    let file = Reason.string_of_source ~strip_root file_key in
    let untainted =
      if trust then
        Some untainted
      else
        None
    in
    (file, untainted, covered, total, trusted_percentage, percentage)
  in
  if json then
    Hh_json.(
      let file_to_json stats =
        let (file, trusted, covered, total, trusted_percentage, percentage) = file_stats stats in
        let percentage = [("percentage", JSON_Number (spf "%0.2f" percentage))] in
        let percentage =
          match trusted_percentage with
          | Some p -> ("trusted_percentage", JSON_Number (spf "%0.2f" p)) :: percentage
          | None -> percentage
        in
        let covered = [("covered", int_ covered)] in
        let covered =
          match trusted with
          | Some t -> ("trusted", int_ t) :: covered
          | None -> covered
        in
        JSON_Object ([("file", string_ file)] @ percentage @ covered @ [("total", int_ total)])
      in
      let array_ elts = JSON_Array elts in
      let file_list =
        Base.List.sort ~compare:(fun (a, _) (b, _) -> Stdlib.compare a b) stats
        |> Base.List.map ~f:file_to_json
      in
      let covered_expressions = [("covered_expressions", int_ covered)] in
      let covered_expressions =
        if trust then
          ("trusted_expressions", int_ trusted) :: covered_expressions
        else
          covered_expressions
      in
      let percentage = [("percentage", JSON_Number (spf "%0.2f" percentage))] in
      let percentage =
        if trust then
          ("trusted_percentage", JSON_Number (spf "%0.2f" trusted_percentage)) :: percentage
        else
          percentage
      in
      let json_output =
        JSON_Object
          [
            ("files", array_ file_list);
            ( "statistics",
              JSON_Object
                ([("files_in_directory", int_ num_files_in_dir)]
                @ covered_expressions
                @ [("total_expressions", int_ total)]
                @ percentage) );
          ]
      in
      print_json_endline ~pretty json_output)
  else
    let (truncation_text, truncated_stats) =
      if num_files_in_dir > 50 && not show_all then
        ( spf
            "\nOnly showing coverage for 50 of %d files. To show more, rerun with --show-all.\n"
            num_files_in_dir,
          Base.List.take stats 50 )
      else
        ("", stats)
    in
    if num_files_in_dir > 0 then (
      print_endlinef "\nCoverage results from %d file(s):\n" num_files_in_dir;
      Base.List.iter
        ~f:(fun fstats ->
          let (file, trusted, covered, total, trusted_percentage, percentage) = file_stats fstats in
          match (trusted_percentage, trusted) with
          | (Some p, Some t) ->
            print_endlinef
              "%s: %0.2f%% trusted (%d of %d expressions), %0.2f%% covered (%d of %d expressions)"
              file
              p
              t
              total
              percentage
              covered
              total
          | _ -> print_endlinef "%s: %0.2f%% (%d of %d expressions)" file percentage covered total)
        truncated_stats;
      print_endline truncation_text
    );

    print_endlinef "-----------------------------------";
    print_endlinef "Aggregate coverage statistics";
    print_endlinef "-----------------------------------";
    print_endlinef "Files                : %d" num_files_in_dir;
    print_endlinef "Expressions          :";
    if trust then print_endlinef "  Trusted            : %d" trusted;
    print_endlinef "  Covered            : %d" covered;
    print_endlinef "  Total              : %d" total;
    if trust then print_endlinef "  Trust Percentage   : %0.2f%%" trusted_percentage;
    print_endlinef "  Covered Percentage : %0.2f%%" percentage;
    print_endlinef ""

let main
    base_flags
    option_values
    json
    pretty
    root
    strip_root
    wait_for_recheck
    input
    show_all
    trust
    files
    () =
  let flowconfig_name = base_flags.Base_flags.flowconfig_name in
  let batch =
    get_filenames_from_input input files |> Base.List.map ~f:(Path.make %> Path.to_string)
  in
  let input = Base.Option.map (Base.List.hd batch) (fun x -> File_input.FileName x) in
  let root = get_the_root ~base_flags ?input root in
  (* pretty implies json *)
  let json = json || pretty in
  let request = ServerProt.Request.BATCH_COVERAGE { batch; wait_for_recheck; trust } in
  match connect_and_make_request flowconfig_name option_values root request with
  | ServerProt.Response.BATCH_COVERAGE (Error msg) -> Exit.(exit ~msg Unknown_error)
  | ServerProt.Response.BATCH_COVERAGE (Ok resp) ->
    output_results ~root ~strip_root ~json ~pretty ~show_all ~trust resp
  | response -> failwith_bad_response ~request ~response

let command = CommandSpec.command spec main

File: rust_port/crates/flow_cli/src/batch_coverage_command.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::path::Path;

use flow_server_env::server_prot;
use flow_services_coverage::FileCoverage;
use serde_json::json;

use crate::command_spec;
use crate::command_spec::arg_spec;
use crate::command_utils;

fn spec() -> command_spec::Spec {
    let exe_name = command_utils::exe_name();
    let spec = command_spec::Spec::new(
        "batch-coverage",
        "Shows aggregate coverage information for a group of files or directories ",
        command_spec::Visibility::Public,
        format!(
            "Usage: {exe_name} batch-coverage [OPTION]... [FILE...] \n\ne.g. {exe_name} batch-coverage foo.js bar.js baz.js dirname1 dirname2 --show-all \nor   {exe_name} batch-coverage --input-file filenames.txt\n"
        ),
    );
    let spec = command_utils::add_base_flags(spec);
    let spec = command_utils::add_connect_flags_no_lazy(spec);
    let spec = command_utils::add_json_flags(spec);
    let spec = command_utils::add_root_flag(spec);
    let spec = command_utils::add_strip_root_flag(spec);
    let spec = command_utils::add_wait_for_recheck_flag(spec);
    spec.flag(
        "--input-file",
        &arg_spec::optional(arg_spec::string()),
        "File containing list of files or directories to compute coverage for, one per line. If -, the list is read from standard input.",
        None,
    )
    .flag(
        "--show-all",
        &arg_spec::truthy(),
        "Whether to output the coverage for all files. If not specified, this command will only print coverage for 50 files.",
        None,
    )
    .anon("FILE...", &arg_spec::list_of(arg_spec::string()))
}

fn output_results(
    root: &Path,
    strip_root: bool,
    json: bool,
    pretty: bool,
    show_all: bool,
    stats: Vec<(flow_parser::file_key::FileKey, FileCoverage)>,
) {
    fn percent(top: i32, bottom: i32) -> f64 {
        if bottom == 0 {
            0.0
        } else {
            top as f64 / bottom as f64 * 100.0
        }
    }

    // Compute aggregate stats
    let (covered, any, empty) = stats.iter().fold(
        (0, 0, 0),
        |(acc_checked, acc_any, acc_empty), (_, coverage)| {
            (
                acc_checked + coverage.checked,
                acc_any + coverage.uncovered,
                acc_empty + coverage.empty,
            )
        },
    );
    let num_files_in_dir = stats.len();
    let total = covered + any + empty;
    let percentage = percent(covered, total);
    let file_stats = |stats: &(flow_parser::file_key::FileKey, FileCoverage)| {
        let (file_key, coverage) = stats;
        let total = coverage.checked + coverage.uncovered + coverage.empty;
        let percentage = percent(coverage.checked, total);
        let file = file_key.to_absolute();
        let file = if strip_root {
            flow_common::files::relative_path(root, &file)
        } else {
            file
        };
        (file, coverage.checked, total, percentage)
    };

    if json {
        let mut sorted_stats = stats.iter().map(file_stats).collect::<Vec<_>>();
        sorted_stats.sort_by(
            |(a, _, _, _): &(String, i32, i32, f64), (b, _, _, _): &(String, i32, i32, f64)| {
                a.cmp(b)
            },
        );
        let file_list: Vec<serde_json::Value> = sorted_stats
            .into_iter()
            .map(
                |(file, covered, total, percentage): (String, i32, i32, f64)| {
                    json!({
                        "file": file,
                        "percentage": format!("{percentage:.2}").parse::<f64>().unwrap(),
                        "covered": covered,
                        "total": total,
                    })
                },
            )
            .collect();
        let json_output = json!({
            "files": file_list,
            "statistics": {
                "files_in_directory": num_files_in_dir,
                "covered_expressions": covered,
                "total_expressions": total,
                "percentage": format!("{percentage:.2}").parse::<f64>().unwrap(),
            },
        });
        flow_hh_json::print_json_endline(pretty, &json_output);
        return;
    }

    let (truncation_text, truncated_stats) = if num_files_in_dir > 50 && !show_all {
        (
            format!(
                "\nOnly showing coverage for 50 of {} files. To show more, rerun with --show-all.\n",
                num_files_in_dir
            ),
            stats.into_iter().take(50).collect::<Vec<_>>(),
        )
    } else {
        (String::new(), stats)
    };

    if num_files_in_dir > 0 {
        println!("\nCoverage results from {} file(s):\n", num_files_in_dir);
        for stats in &truncated_stats {
            let (file, covered, total, percentage) = file_stats(stats);
            println!("{file}: {percentage:.2}% ({covered} of {total} expressions)");
        }
        println!("{truncation_text}");
    }

    println!("-----------------------------------");
    println!("Aggregate coverage statistics");
    println!("-----------------------------------");
    println!("Files                : {}", num_files_in_dir);
    println!("Expressions          :");
    println!("  Covered            : {}", covered);
    println!("  Total              : {}", total);
    println!("  Covered Percentage : {percentage:.2}%");
    println!();
}

fn main(args: &arg_spec::Values) {
    let base_flags = command_utils::get_base_flags(args);
    let flowconfig_name = base_flags.flowconfig_name;
    let connect_flags = command_utils::get_connect_flags(args);
    let json_flags = command_utils::get_json_flags(args);
    let root_arg =
        command_spec::get(args, "--root", &arg_spec::optional(arg_spec::string())).unwrap();
    let strip_root = command_spec::get(args, "--strip-root", &arg_spec::truthy()).unwrap();
    let wait_for_recheck = command_spec::get(
        args,
        "--wait-for-recheck",
        &arg_spec::optional(arg_spec::bool_flag()),
    )
    .unwrap();
    let input_file = command_spec::get(
        args,
        "--input-file",
        &arg_spec::optional(arg_spec::string()),
    )
    .unwrap();
    let show_all = command_spec::get(args, "--show-all", &arg_spec::truthy()).unwrap();
    let files = command_spec::get(args, "FILE...", &arg_spec::list_of(arg_spec::string()))
        .unwrap()
        .unwrap_or_default();

    let batch = command_utils::get_filenames_from_input(false, input_file.as_deref(), Some(&files))
        .into_iter()
        .map(|filename| command_utils::expand_path(&filename))
        .collect::<Vec<_>>();
    let input = batch
        .first()
        .map(|filename| flow_server_utils::file_input::FileInput::FileName(filename.clone()));
    let root = command_utils::get_the_root(&flowconfig_name, root_arg.as_deref(), input.as_ref());
    // pretty implies json
    let json = json_flags.json || json_flags.pretty;
    let request = server_prot::request::Command::BATCH_COVERAGE {
        batch,
        wait_for_recheck,
    };
    let response =
        command_utils::connect_and_make_request(&flowconfig_name, &connect_flags, &root, &request);
    match response {
        server_prot::response::Response::BATCH_COVERAGE(Err(msg)) => {
            eprintln!("{}", msg);
            flow_common_exit_status::exit(flow_common_exit_status::FlowExitStatus::UnknownError);
        }
        server_prot::response::Response::BATCH_COVERAGE(Ok(response)) => {
            output_results(
                &root,
                strip_root,
                json,
                json_flags.pretty,
                show_all,
                response,
            );
        }
        response => {
            command_utils::failwith_bad_response(&request, &response);
        }
    }
}

pub(crate) fn command() -> command_spec::Command {
    command_spec::command(spec(), main)
}
